react/features/videosipgw/middleware.js

```diff
@@ -0,0 +1,182 @@
+    case SIP_GW_INVITE_ROOMS: {
+        const { status } = getState()['features/videosipgw'];
+
+        if (status === JitsiSIPVideoGWStatus.STATUS_UNDEFINED) {
+            dispatch(showErrorNotification({
+                descriptionKey: 'recording.unavailable',
+                descriptionArguments: {
+                    serviceName: '$t(videoSIPGW.serviceName)'
+                },
+                titleKey: 'videoSIPGW.unavailableTitle'
+            }));
+
+            return;
+        } else if (status === JitsiSIPVideoGWStatus.STATUS_BUSY) {
+            dispatch(showWarningNotification({
+                descriptionKey: 'videoSIPGW.busy',
+                titleKey: 'videoSIPGW.busyTitle'
+            }));
+
+            return;
+        } else if (status !== JitsiSIPVideoGWStatus.STATUS_AVAILABLE) {
+            logger.error(`Unknown sip videogw status ${status}`);
+
+            return;
+        }
+
+        for (const room of action.rooms) {
+            const { id: sipAddress, name: displayName } = room;
+
+            if (sipAddress && displayName) {
+                const newSession = action.conference
+                    .createVideoSIPGWSession(sipAddress, displayName);
+
+                if (newSession instanceof Error) {
+                    const e = newSession;
+
+                    if (e) {
+                        switch (e.message) {
+                        case JitsiSIPVideoGWStatus.ERROR_NO_CONNECTION: {
+                            dispatch(showErrorNotification({
+                                descriptionKey: 'videoSIPGW.errorInvite',
+                                titleKey: 'videoSIPGW.errorInviteTitle'
+                            }));
+
+                            return;
+                        }
+                        case JitsiSIPVideoGWStatus.ERROR_SESSION_EXISTS: {
+                            dispatch(showWarningNotification({
+                                titleKey: 'videoSIPGW.errorAlreadyInvited',
+                                titleArguments: { displayName }
+                            }));
+
+                            return;
+                        }
+                        }
+                    }
+                    logger.error(
+                        'Unknown error trying to create sip videogw session',
+                        e);
+
+                    return;
+                }
+
+                newSession.start();
+            } else {
+                logger.error(`No display name or sip number for ${
+                    JSON.stringify(room)}`);
+            }
+        }
+    }
+    }
+
+    return result;
+});
+
+/**
+ * Signals that sip gw availability had changed.
+ *
+ * @param {string} status - The new status of the service.
+ * @returns {{
+ *     type: SIP_GW_AVAILABILITY_CHANGED,
+ *     status: string
+ * }}
+ * @private
+ */
+function _availabilityChanged(status: string) {
+    return {
+        type: SIP_GW_AVAILABILITY_CHANGED,
+        status
+    };
+}
```

Comment:
It's okay style-wise to have actions inside the middleware if they're not intended to be used publicly?

Reply:
The _availabilityChanged is private. And the intention is just to notify the reducer to store it in the state for future reference (showing notifications, handled in SIP_GW_INVITE_ROOMS case up in the code).

Reply:
Right, what I'm asking is in terms of style/organization, normally all actions go into an actions.js file but this is a special case because the action is meant to be private. So I want to ask specifically if you've given thought about if it's okay or not to have an action in a middleware file. If you've given it thought and it's fine then I'm okay with it (and someone else can just change it when they want).